### PR TITLE
Add Python 3.10 support

### PR DESCRIPTION
**Pull request by ancalita**
**Proposed changes**:
- Update Python constraint in `pyproject.toml` and then lock file
- Amend CI workflows
- Fixes [failing model training unit tests](https://github.com/RasaHQ/rasa/actions/runs/3497836220/jobs/5857435077) by decoding the temporary directory which is of type `bytes` in Python 3.10 `tempfile` module.
- Filter [3.10 deprecation warnings](https://docs.python.org/3.10/whatsnew/3.10.html#deprecated) from [failing unit tests](https://github.com/RasaHQ/rasa/actions/runs/3522274827/jobs/5905070529).

**Status (please check what you already did)**:
- [ ] added some tests for the functionality
- [x] updated the documentation
- [x] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/main/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

File: docs/docs/installation/installing-rasa-open-source.mdx
Comment (ancalita) [inline docs/docs/installation/installing-rasa-open-source.mdx:87]: I couldn't find anywhere in the docs where we specify supported Python versions 🤔

File: docs/docs/installation/installing-rasa-open-source.mdx
Comment (m-vdb) [inline docs/docs/installation/installing-rasa-open-source.mdx:92]: maybe worth adding to the list of [ignored words](https://github.com/RasaHQ/rasa/blob/89671f0847009f6c2d34c09a919b53e6219b11fc/.typo-ci.yml#L64)

File: pyproject.toml
Comment (m-vdb) [inline pyproject.toml:104]: can we release an alpha for the SDK?

File: rasa/core/training/interactive.py
Comment (m-vdb) [inline rasa/core/training/interactive.py:1727]: 🤔 curious why this change is introduced

File: rasa/engine/caching.py
Comment (m-vdb) [inline rasa/engine/caching.py:309]: 🤔 weird, I can't find info about this behaviour change in the release notes

File: rasa/utils/plotting.py
Comment (m-vdb) [inline rasa/utils/plotting.py:62]: have you tried using the new [`ParamSpec`](https://docs.python.org/3/library/typing.html#typing.ParamSpec)?

File: tests/examples/test_example_bots_training_data.py
Comment (m-vdb) [inline tests/examples/test_example_bots_training_data.py:34]: shouldn't it be "experimental"?

File: tests/conftest.py
Comment (m-vdb) [inline tests/conftest.py:797]: not sure why we're deep-copying the list, as we're just iterating on it, and we anyway pop from the `records` object at line 804

File: docs/docs/installation/installing-rasa-open-source.mdx
Comment (m-vdb) [reply docs/docs/installation/installing-rasa-open-source.mdx:87]: I remember a discussion about it not being clearly explained in the docs

File: pyproject.toml
Comment (ancalita) [reply pyproject.toml:104]: As soon as we merge the rasa-sdk PR🤞🏻 @sanchariGr is it ok to release the sdk alpha earlier than scheduled on 28th Nov?

File: rasa/core/training/interactive.py
Comment (ancalita) [reply rasa/core/training/interactive.py:1727]: Glad you asked this question, as I wasn't sure if this is the right approach.
This unit test was failing with a off by 1 error on the last test case:

```
FAILED tests/core/training/test_interactive.py::test_calc_true_wrapping_width - AssertionError: assert 59 == 60
 +  where 59 = <function calc_true_wrapping_width at 0x7efd101635b0>('牙龈出血的症状对应得疾病可能有：肥大性龈炎、骨髓增生异常综合征、口腔疾病、小儿出血性疾病、边缘性龈炎、获得性维生素K依赖性凝血因子异常、小儿白血病、回归热、坏死性龈口炎、青春期功能失调性子宫出血、郎-奥韦综合征、急性淋巴细胞白血病、感染性血小板减少性紫癜、单纯性牙周炎、单核细胞白血病、δ-贮存池病、小儿特发性血小板减少性紫癜、老年人真性红细胞增多症、急性根尖牙周炎、慢性根尖牙周炎、创伤性口炎、青少年牙周炎、慢性牙周炎', 119)
 +    where <function calc_true_wrapping_width at 0x7efd101635b0> = interactive.calc_true_wrapping_width
```
When I looked into the function internals more closely, I discovered it loops one level lower because in the [iterable](https://github.com/RasaHQ/rasa/blob/89671f0847009f6c2d34c09a919b53e6219b11fc/rasa/core/training/interactive.py#L1725) that gets used in `all()` function, there was an element (120) which fails the condition set by the test to be <= 119. So here I had the option a) change the unit test `wrapping_width` parameter, or b) modify the actual function. In hindsight, probably option a) would disturb the least the codebase. 
I couldn't  figure out why this test case only fails deterministically with Python 3.10, the libraries that get used in this function have not changed at all. Potentially something to do with the decoding of the text?

Comment (ancalita) [reply rasa/core/training/interactive.py:1727]: I also don't know how the test parameters were chosen, perhaps @twerkmeister knows more as you seem to have committed this change?

File: tests/examples/test_example_bots_training_data.py
Comment (ancalita) [reply tests/examples/test_example_bots_training_data.py:34]: Indeed, I'll correct the incorrect word.

File: tests/conftest.py
Comment (ancalita) [reply tests/conftest.py:797]: If I didn't deepcopy the list, and iterated over the original list, I'd be popping elements from the list I'm iterating over, and the looping would finish before all the warnings that need to be removed are removed, so `records` would still contain unfiltered `DeprecationWarnings`.

File: rasa/engine/caching.py
Comment (ancalita) [reply rasa/engine/caching.py:309]: I know 😬 The `mkdtemp()` function returns type AnyStr, which includes bytes, not sure if that's been changed in 3.10.

File: rasa/core/training/interactive.py
Comment (m-vdb) [reply rasa/core/training/interactive.py:1727]: > Potentially something to do with the decoding of the text?

I would assume so, but can find any trace of it in the release note either. Sounds good

File: rasa/engine/storage/local_model_storage.py
Comment (ancalita) [reply rasa/engine/storage/local_model_storage.py:163]: I only added the fixes without making changes to the existing state of the codebase, not sure why there are ways to use a temp directory.
I can definitely look into using the custom class to replace the direct usages of `tempfile` in the files i had to amend in this PR, but will need to create a follow-up ticket to encompass many other files that use `tempfile.TemporaryDirectory()` context manager.

File: rasa/utils/plotting.py
Comment (m-vdb) [reply rasa/utils/plotting.py:62]: > expected to see at least one of the function arguments typed to the same type as the return type

this is weird though - sounds like a mypy error

File: tests/conftest.py
Comment (m-vdb) [reply tests/conftest.py:797]: 🤦🏻 yes, indeed 💤

File: rasa/utils/plotting.py
Comment (ancalita) [reply rasa/utils/plotting.py:62]: Should I mark with a type ignore comment?

File: pyproject.toml
Comment (sanchariGr) [reply pyproject.toml:104]: Would you prefer it be released as soon as the PR is merged? Sure we can do that 👍

File: docs/docs/installation/installing-rasa-open-source.mdx
Comment (ancalita) [reply docs/docs/installation/installing-rasa-open-source.mdx:87]: Addressed: https://github.com/RasaHQ/rasa/pull/11773/commits/bfd5135e0d7127da726a92b50602c29ae1991636

File: docs/docs/installation/installing-rasa-open-source.mdx
Comment (ancalita) [reply docs/docs/installation/installing-rasa-open-source.mdx]: Addressed: https://github.com/RasaHQ/rasa/pull/11773/commits/bfd5135e0d7127da726a92b50602c29ae1991636

File: docs/docs/installation/installing-rasa-open-source.mdx
Comment (ancalita) [reply docs/docs/installation/installing-rasa-open-source.mdx:92]: Addressed: https://github.com/RasaHQ/rasa/pull/11773/commits/bfd5135e0d7127da726a92b50602c29ae1991636

File: rasa/utils/plotting.py
Comment (ancalita) [reply rasa/utils/plotting.py]: Addressed: https://github.com/RasaHQ/rasa/pull/11773/commits/bfd5135e0d7127da726a92b50602c29ae1991636

File: tests/examples/test_example_bots_training_data.py
Comment (ancalita) [reply tests/examples/test_example_bots_training_data.py]: Addressed: https://github.com/RasaHQ/rasa/pull/11773/commits/bfd5135e0d7127da726a92b50602c29ae1991636

File: rasa/engine/storage/local_model_storage.py
Comment (ancalita) [reply rasa/engine/storage/local_model_storage.py]: Addressed: https://github.com/RasaHQ/rasa/pull/11773/commits/bfd5135e0d7127da726a92b50602c29ae1991636 and created a new task ticket in Atom backlog (ATO-560) for the rest of the modules.

**Review by tmbo — COMMENTED** [review]
lgtm :+1: great work, digging into the dependency updates.

do we also want to update the python version used in our docker containers or should that be separate?

**Comment by m-vdb** [comment]
> do we also want to update the python version used in our docker containers or should that be separate?

@tmbo I believe it's separate work. Infra will take care of that: they needed to upgrade the OS of docker images to address vulnerabilities, but the upgrade was blocked because the next ubuntu version pins 3.10. So I think it's going to be done as part of that work. ([this ticket](https://rasahq.atlassian.net/browse/INFRA-195)). I will double check with Infra to make sure this is done. Thanks for the nudge

**Comment by m-vdb** [comment]
@ancalita one dead link in the docs tests makes the check fail, but it appears the link is momentarily dead, nothing we can do about it straight away. if you can merge the PR when you're ready, LMK and I'll merge even if the check is still red

**Comment by ancalita** [comment]
> @ancalita one dead link in the docs tests makes the check fail, but it appears the link is momentarily dead, nothing we can do about it straight away. if you can merge the PR when you're ready, LMK and I'll merge even if the check is still red

@m-vdb Yup sure, I'm now pending two things actually:
- rasa-sdk alpha release
- feedback from @twerkmeister on that rasa interactive width calculation function

**Comment by twerkmeister** [comment]
@ancalita I am not sure what the width calculation refers to - can you give me some pointer at what to look at? I can't see any specific question about it

**Comment by ancalita** [comment]
> @ancalita I am not sure what the width calculation refers to - can you give me some pointer at what to look at? I can't see any specific question about it

Sure @twerkmeister Here is the [calc_true_wrapping_width() code](https://github.com/RasaHQ/rasa/blob/89671f0847009f6c2d34c09a919b53e6219b11fc/rasa/core/training/interactive.py#L1698), this is the [unit test](https://github.com/RasaHQ/rasa/blob/89671f0847009f6c2d34c09a919b53e6219b11fc/tests/core/training/test_interactive.py#L820), the last test case fails with `AssertionError: 59 != 60` only with Python 3.10.

File: pyproject.toml
Comment (ancalita) [reply pyproject.toml]: Updated to `3.4`: https://github.com/RasaHQ/rasa/pull/11773/commits/94dc6610ddfebd797ff87170b709847a9c57af6d

File: rasa/core/training/interactive.py
Comment (twerkmeister) [reply rasa/core/training/interactive.py:1727]: Hmmm The context of this is the following ~ the rasa interactive console uses some table library that crashes when  text overflows. This does not happen when only dealing with latin alphabet. However many character sets in unicode such as Chinese characters are wider than a character in the latin alphabet when printed in console and the table layout does not account for that. That's why you need to calculate their width separately. 

What might have happened here is that the width of one or some of those characters changed in the unicode package of 3.10 🤔 making it wider and thus we have to wrap at 59 characters instead of 60 now for that specific string when the terminal width is 119. I think it's fine to increase the theoretical terminal width in the test to 120 and keep the target wrapping width at 60 if that then works in all supported python versions alike. Since the py310 change seems to make the result more conservative (wrapping already at 59 characters instead of 60) I'd say let's just roll with it and not worry too much.

**Review by m-vdb — APPROVED** [review]
⭐ ⭐ ⭐

**Pull request merged**
